File: Day272/eventualSafeStates.go
package main

import (
	"fmt"
	"sort"
)

func create(n int, mat [][]int) [][]int {
	adj := make([][]int, 0)
	for i:=0; i<n; i++ {
		adj = append(adj, make([]int, 0))
	}
	for _,v := range mat {
		adj[v[0]] = append(adj[v[0]], v[1])
	}
	return adj
}

func eventualSafeNodes(n int, adj [][]int) []int {
	rev := make([][]int, n)
	for i:=0; i<n; i++ {
		rev[i] = make([]int, 0)
	}
	id := make([]int, n)
	for i:=0; i<n; i++ {
		for _,v := range adj[i] {
			rev[v] = append(rev[v], i)
			id[i]++
		}
	}
	queue := make([]int, 0)
	answer := make([]int, 0)
	for i:=0; i<n; i++ {
		if id[i] == 0 {
			queue = append(queue, i)
		}
	}
	for len(queue) > 0 {
		temp := queue[0]
		queue = queue[1:]
		answer = append(answer, temp);
		for _,n := range rev[temp] {
			id[n]--
			if id[n] == 0 {
				queue = append(queue, n)
			}
		}
		sort.Ints(answer)
	}
	return answer
}


func main() {
	n1 := 7;
	mat1 := [][]int{ {0,1},{0,2}, {1,2}, {1,3}, {2,5}, {3,0}, {4,5}}
	n2 := 4;
	mat2 := [][]int{{0,1}, {1,2}, {2,0}, {2,3}}
	adj1 := create(n1, mat1)
	adj2 := create(n2, mat2)
	// fmt.Println(adj1)
	// fmt.Println(adj2)
	fmt.Println(eventualSafeNodes(n1, adj1))
	fmt.Println(eventualSafeNodes(n2, adj2))
}

File: Day272/eventualSafeStates.java
package Day272;

import java.util.ArrayList;
import java.util.List;

import Day254.booleanMatrix;

public class eventualSafeStates {
    public static List<List<Integer>> create(int n, int mat[][]) {
        List<List<Integer>> answer = new ArrayList<>();
        for (int i=0; i<n; i++) {
            answer.add(new ArrayList<>());
        }
        for (int i=0; i<mat.length; i++) {
            answer.get(mat[i][0]).add(mat[i][1]);
        }
        return answer;
    }
    public static List<Integer> ESS(int V, List<List<Integer>> adj) {
        List<Integer> answer =  new ArrayList<>();
        boolean visited[] = new boolean[V];
        boolean path[] = new boolean[V];
        boolean temp[] = new boolean[V];
        for (int i=0; i<V; i++) {
            if (!visited[i]) {
                dfs(i, adj, visited, path, temp);
            }
        }
        for (int i=0; i<V; i++) {
            if (temp[i]) {
                answer.add(i);
            }
        }
        return answer;
    }

    private static boolean dfs(int i ,List<List<Integer>> adj, boolean visited[], boolean path[], boolean temp[]) {
        visited[i] = true;
        path[i] = true;
        temp[i] = false;
        for (int n : adj.get(i)) {
            if (!visited[n]) {
                if (dfs(n, adj, visited, path, temp)) {
                    return true;
                }
            } else if (path[n]) {
                return true;
            }
        }
        temp[i] = true;
        path[i] = false;
        return false;
    }
    public static void main(String[] args) {
        int n1 = 7;
        int mat1[][] = { {0,1},{0,2}, {1,2}, {1,3}, {2,5}, {3,0}, {4,5}};
        int n2 = 4;
        int mat2[][] = {{0,1}, {1,2}, {2,0}, {2,3}};
        List<List<Integer>> adj1 = create(n1, mat1);
        List<List<Integer>> adj2 = create(n2, mat2);
        System.out.println(ESS(n1, adj1));
        System.out.println(ESS(n2, adj2));
    }
}